readme.txt
=== Ban Hammer ===
Contributors: Ipstenu
Tags: email, ban, registration, buddypress, wpmu, multisite
Requires at least: 5.5
Tested up to: 6.0
Stable tag: 2.8
Donate Link: https://ko-fi.com/A236CEN/
License: GPLv2

Prevent people from registering via your disallowed comment keys.

== Description ==

We've all had this problem: a group of spammers from `mail.ru` are registering to your blog, but you want to keep registration open.  How do you kill the spammers without bothering your clientele?  While you could edit your theme's `functions.php` and block the domain, once you get past a few bad eggs, you have to escalate.

Ban Hammer helps you do that by preventing unwanted users from registering.

New in version 2.9: Tired of people registering with user name 'manager', 'admin', 'administrator', or 'owner'? Ban Hammer now does the same check on user name/login entry as it's always done for the email address. See the 'User Name Hammer' section below for details.

On a single install of WordPress, instead of using its own database table, Ban Hammer pulls from your list of prohibited emails from the Disallowed Comment Keys feature, native to WordPress.  Since emails never equal IP addresses, it simply skips over and ignores them.

On a network instance, there's a network wide setting for banned emails and domains. This means you only have <em>one</em> place to update and maintain your blocked list.  When a listed user attempts to register, they get a customizable message that they cannot register.

For advanced documentation, including how to use on WooCommerce, please visit [the Ban Hammer Wiki](https://github.com/Ipstenu/ban-hammer/wiki).

* [Development](https://github.com/Ipstenu/ban-hammer/)
* [Wiki](https://github.com/Ipstenu/ban-hammer/wiki)
* [Donate](https://ko-fi.com/A236CEN/)

= Privacy Policy =

This plugin does not track data outside of what WordPress already collects. It utilizes the submitted email address to validate the domain and compares it to the list of prohibited domains and emails. No additional data is processed.

= Credits =

Ban Hammer is a very weird fork of [Philippe Paquet's No Disposable Email plugin](http://www.joeswebtools.com/wordpress-plugins/no-disposable-email/). The original plugin was a straight forward .dat file that listed all the bad emails (generally ones like mailinator that are disposable) and while Ban Hammer doesn't do that, this would not have been possible without that which was done before.

Many thanks are due to WP-Deadbolt, for making me think about SQL and TTC for StopForumSpam integration. MASSIVE credit to Travis Hamera for the StopForumSpam/cURL fix! And then props to Helen Hou-Sandí for not using curl at all. Protip? Use <a href="http://planetozh.com/blog/2009/08/how-to-make-http-requests-with-wordpress/">WP_http</a> instead!

==Changelog==

* June 2022 by Starbuck
* Add support for testing user name in addition to existing user email test

== Installation ==

<strong>Single Site (Traditional) Install</strong>

After installation, go to **Tools > Ban Hammer** to customize the error message (and banned emails, but it's the same list from your comment moderation so...).

<strong>Multisite</strong>

After installation, go to **Network Admin > Settings > Ban Hammer** to customize the error message and banned email list. This will ban users network wide.

== Screenshots ==

1. Default Error message
2. Admin screen
3. Ban Hammer Users
4. Users Menu, with Spammer Flag on
5. BuddyPress Error message

== User Name Hammer ==

By default, Ban Hammer tests all email addresses entered during registration. Activating the same testing for the user name is as simple as checking a box and saving the configuration. Remember that the ban words are tested as substrings. So as noted below, if the ban text is 'cookie' and 'cookiemonster' from Sesame Street won't be able to create an account. :(

Email address testing is always done - nothing changes there. This new feature only adds more functionality. At this time email address testing cannot be disabled.

== Frequently Asked Questions ==

= If I change the Blocklist via Ban Hammer, will it change the Disallowed Comment Keys? =

On **single site installs**, yes. They are the exact same list, they use the same fields and they update the same data. The only reason I put it there was I felt having an all-in-one place to get the data would be better.

= Does this list the rejected registers? =

No. Since WordPress itself doesn't list rejected comments, the rejected users are similarly lost forever.

= Where did Stop Forum Spam go? =

This plugin no longer uses Stop Forum Spam. If you need that feature, please use <a href="http://wordpress.org/extend/plugins/stop-spammer-registrations-plugin/">Stop Spammer Registrations</a> instead. They did it way better.

= Does this work on MultiSite? =

Yes it does, but a little differently. If you're using multisite, instead of pulling from the Disallowed Comment Keys (which is per site), you have a separate list off Network Admin -> Settings. This is because you only want to have the network admins determining who can register on the network.

= Does this work on BuddyPress? =

I believe so.

= Does this work on WooCommerce? =

You have to make your own hook because WooCommerce doesn't use the normal registration functions. Don't panic. I have [directions here](https://github.com/Ipstenu/ban-hammer/wiki#woocommerce).

= Can I block partials? =

Yes but not wildcards. If you put in `viagra` for example, you will block `viagrajones@gmail.com` _and_ `john@viagra.com` so please use this carefully. If you put in `cookie` then you'll block `cookiemonster@sesamestreet.edu` and everyone would be sad.

Similarly, registration will be blocked if you use 'cookie', you are checking the user login, and the name is 'morecookiesplease'.

If you want to block everyone from all subdomains (like `joe@bar.example.com`) then you can block `.example.com` and that will block all the subdomains.


changelog.txt
= 2.9.0 =
* June 2022 by Starbuck
* Refactor Drop testing to support more than one field
* Add support for testing user name in addition to existing user email test

= 2.8 =
* February 2021 by Ipstenu
* Removing check for if registration is active

= 2.7.1 =
* August 2020 by Ipstenu
* Typo fixed. Saving works. My bad!

= 2.7 =
* August 2020 by Ipstenu
* Compatibility with 5.5
* Corrected deprecation of blacklist_keys (it still falls back for older WP)

= 2.6.2 =
* December 2018 by Ipstenu
* PHPCS checks
* Check for WooCommerce

= 2.6.1 =
* January 2018 by Ipstenu
* Formatting changes
* Stopped non-admins from seeing settings links and notifications

= 2.6 =
* February 2017 by Ipstenu
* Allow redirection to custom URLs on failed login.
* Move plugin to Settings API
* Combine options
* Fixed BuddyPress
* Optimized multisite
* Removed check for WP 3.4 (only 4.0 and up get updates anyway)

= 2.5.4 =
* 27 April 2015 by Ipstenu
* Changing priority to take precedence over other plugins that set new users to require approval.

= 2.5.2 & 2.5.3 =
* 17 November 2015 by Ipstenu
* Rolling back BuddyPress changes to correct

= 2.5.1 =
* 15 November 2015 by Ipstenu
* Bad sanitization choice killed settings for some people.

= 2.5.1 =
* 13 November 2015 by Ipstenu
* People on Multisite couldn't access settings. There's a level of irony there.

= 2.5 =
* 12 November 2015 by Ipstenu
* Update for WP 4.4
* Code moved to a singleton
* Improved sanitization of emails
* Fixed link to settings page

= 2.4.1 =
* 6 August 2015 by Ipstenu
* Security filtering improvements

= 2.4 =
* 1 October 2014 by Ipstenu
* Fixing languages
* Fixing define notice due to messing up text domain.

= 2.3 =
* 29 October 2013 by Ipstenu
* Languages for 3.7
* Compat for 3.7

= 2.2 =
* 15 January 2013 by Ipstenu
* Fixed translation bug
* Dropping support for older versions

= 2.1 =
* 26 August 2012 by Ipstenu
* The return of Multisite
* BuddyPress improvements
* Dropping support for older versions

= 2.0 =
* 30 May 2012 by Ipstenu
* Removal of Stop Forum Spam.
* Whole plugin was consolidated to run faster, with fewer files, and with contextual help. Egad, I've learned a lot since 2009!

= 1.7 =
* 24 April 2012 by Ipstenu
* Proper uninstallation.

= 1.6.1 =
* 17 April 2012 by Ipstenu
* Cleanup. Nothing major here, just documentation and all.

= 1.6 =
* 05 August 2011 by Ipstenu
* Internationalization.

= 1.5.2 =
* 09 March 2011 by Ipstenu
* Bugfix.  Typo made it NOT enableable.

= 1.5 =
* 08 March 2011 by Ipstenu
* Allows for deletion of spammers from the User List (credit mario_7)
* Added optional functionality to show spammer status on the normal users list.
* Moved Ban Hammer Users to the USERS menu (now called 'Ban Hammered')
* Works on BuddyPress!

= 1.4 =
* 16 August 2010 by Ipstenu
* Checks for presence of the cURL extension. If not found, the option to use StopForumSpam is removed. (using http://cleverwp.com/function-curl-php-extension-loaded/ as recommended by kmaisch )

= 1.3 =
* 08 July 2010  by Ipstenu
* Pulling out the WPMU stuff that's never going to happen now that it's MultiSite and doesn't work.

= 1.2 =
* 08 November 2009  by Ipstenu
* This lists all users marked by StopForumSpam as spammers, if you're using that option (and not if not). (Thanks to obruchez for the suggestion!).

= 1.1 =
* 03 May 2009 by Ipstenu
* Subversion before coffee = BAD.

=  1.0 =
* 03 May 2009 by Ipstenu
* First public version.

=  0.3 =
* 30 March 2009 by Ipstenu
* The error message is customizable.
* Added support for StopForumSpam.com
* Added in checkbox to use StopForumSpam (default to NO).
* Cleans up after itself on deactivation (deletes the banhammer_foo values from the wp_options table because I HATE when plugins leave themselves).

=  0.2 =
* 29 March 2009 by Ipstenu
* Shifted to use the WordPress comment blacklist as source. This was pretty much an 80% re-write from NDE's basis, keeping only the basic check at registration code.

=  0.1 =
* 28 March 2009 by Ipstenu
* First release using No Disposable Email's .dat file as a source.


ban-hammer.php
<?php
/*
Plugin Name: Ban Hammer
Plugin URI: http://halfelf.org/plugins/ban-hammer/
Description: Prevent people from registering with any email or user name you list.
Version: 2.9
Author: Mika Epstein
Author URI: http://halfelf.org/
Network: true
Text Domain: ban-hammer

Copyright 2009-21 Mika Epstein (email: ipstenu@halfelf.org)

	This file is part of Ban Hammer, a plugin for WordPress.

	Ban Hammer is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 2 of the License, or
	(at your option) any later version.

	Ban Hammer is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with WordPress.  If not, see <http://www.gnu.org/licenses/>.

*/

class BanHammer {

	/*
	 * Starter defines and vars for use later
	 *
	 * @since 2.6
	 */

	// Holds option data.
	public $option_name = 'banhammer_options';
	public $option_defaults;
	public $options;

	// DB version, for schema upgrades.
	public $db_version = 2;

	// Constants
	public $buddypress;
	public $bannedlist;

	// Instance
	public static $instance;

	/**
	 * Construct
	 *
	 * @since 2.5
	 * @access public
	 */
	public function __construct() {

		//allow this instance to be called from outside the class
		self::$instance = $this;

		add_action( 'init', array( &$this, 'init' ) );
		add_action( 'admin_init', array( &$this, 'admin_init' ) );

		//add admin panel
		if ( is_multisite() ) {
			add_action( 'network_admin_menu', array( &$this, 'network_admin_menu' ) );
			add_action( 'current_screen', array( &$this, 'network_admin_screen' ) );
		} else {
			add_action( 'admin_menu', array( &$this, 'admin_menu' ) );
		}

		// Setting plugin defaults here:
		$this->option_defaults = array(
			'db_version'   => $this->db_version,
			'redirect'     => 'no',
			'redirect_url' => 'http://example.com',
			'message'      => __( '<strong>ERROR</strong>: Your email has been banned from registration.', 'ban-hammer' ),
			'check_user_login' => false
		);

		// Fetch and set up options.
		$this->options = wp_parse_args( get_site_option( $this->option_name ), $this->option_defaults, false );

		if ( is_multisite() ) {
			$this->bannedlist = get_site_option( 'banhammer_keys', 'spammer@example.com', false );
		} elseif ( false !== get_option( 'disallowed_keys' ) ) {
			$this->bannedlist = get_option( 'disallowed_keys' );
		} else {
			$this->bannedlist = get_option( 'blacklist_keys' );
		}

		// check if DB needs to be upgraded (this will merge old settings to new)
		$naked_options = get_site_option( $this->option_name );

		if ( ! isset( $naked_options['db_version'] ) || $naked_options['db_version'] < $this->db_version ) {
			if ( isset( $old_b_message ) && ! is_null( $old_b_message ) ) {
				$current_db_version = 0;
			} else {
				$current_db_version = isset( $naked_options['db_version'] ) ? $naked_options['db_version'] : 0;
			}
			$this->upgrade( $current_db_version );
		}
	}

	/**
	 * Init
	 *
	 * @since 2.5
	 * @access public
	 */
	public function init() {
		// Filter for if multisite but NOT BuddyPress
		if ( is_multisite() && ! defined( 'BP_PLUGIN_DIR' ) ) {
			add_filter( 'wpmu_validate_user_signup', array( &$this, 'wpmu_validation' ), 99 );
		}

		if ( defined( 'BP_PLUGIN_DIR' ) ) {
			add_filter( 'bp_core_validate_user_signup', array( &$this, 'buddypress_signup' ) );
		}

		// The magic sauce
		add_action( 'register_post', array( &$this, 'banhammer_drop' ), 1, 3 );
	}

	/**
	 * BuddyPress Signup Filter
	 *
	 * @since 2.6
	 * @access public
	 */
	public function buddypress_signup( $result ) {
		if ( $this->banhammer_drop( $result['user_name'], $result['user_email'], $result['errors'] ) ) {
			$result['errors']->add( 'user_email', $this->options['message'] );
		}
		return $result;
	}

	/**
	 * Admin init Callback
	 *
	 * @since 2.6
	 */
	public function admin_init() {

		// Settings links
		add_filter( 'plugin_row_meta', array( &$this, 'donate_link' ), 10, 2 );
		$plugin = plugin_basename( __FILE__ );
		add_filter( 'plugin_action_links_$plugin', array( &$this, 'settings_link' ) );

		// Register Settings
		$this->register_settings();

		// Is Registration Active?
		// Currently disabled to see if this increases support.
		// $this->is_registration_active();
	}

	/**
	 * Admin Notices
	 *
	 * Display admin notices as needed.
	 *
	 * @since 2.6
	 */
	public function admin_notices() {
		printf( '<div class="notice notice-%1$s"><p>%2$s</p></div>', esc_attr( $this->notice_class ), wp_kses_post( $this->notice_message ) );
	}

	/**
	 * Upgrades Database
	 *
	 * @param int $current_db_version the current DB version
	 * @since 2.6
	 */
	public function upgrade( $current_db_version ) {
		if ( $current_db_version < 1 ) {

			// Migrate old options to new
			$this->options['redirect']     = $this->option_defaults['redirect'];
			$this->options['redirect_url'] = $this->option_defaults['redirect_url'];
			$this->options['message']      = get_option( 'banhammer_message', $this->option_defaults['message'] );
			$this->options['db_version']   = '1';

			// Delete old options
			delete_site_option( 'banhammer_message' );
		}
		if ( $current_db_version < 2 ) {
			$this->options['check_user_login']   = false;
			$this->options['db_version']   = '2';
		}

		update_site_option( $this->option_name, $this->options );
	}

	/**
	 * Admin Menu
	 *
	 * @since 2.5
	 * @access public
	 */
	public function admin_menu() {
		add_management_page( __( 'Ban Hammer', 'ban-hammer' ), __( 'Ban Hammer', 'ban-hammer' ), 'moderate_comments', 'ban-hammer', array( &$this, 'options' ) );
	}

	/**
	 * Network Admin Menu
	 *
	 * @since 2.5.1
	 * @access public
	 */
	public function network_admin_menu() {
		add_submenu_page( 'settings.php', __( 'Ban Hammer', 'ban-hammer' ), __( 'Ban Hammer', 'ban-hammer' ), 'manage_networks', 'ban-hammer', array( &$this, 'options' ) );
	}

	/**
	 * Network Admin Screen Callback
	 *
	 * @since 3.0
	 */

	public function network_admin_screen() {
		$current_screen = get_current_screen();
		if ( 'settings_page_ban-hammer-network' === $current_screen->id ) {

			if ( isset( $_POST['update'] ) && check_admin_referer( 'banhammer_networksave' ) ) {
				$options = $this->options;
				$input   = $_POST['banhammer_options']; // phpcs:ignore - Sanitized further down.

				// This is hardcoded for a reason.
				$output['db_version'] = $this->db_version;

				// Message
				if ( $input['message'] !== $options['message'] && wp_kses_post( $input['message'] ) === $input['message'] ) {
					$output['message'] = wp_kses_post( $input['message'] );
				} else {
					$output['message'] = $options['message'];
				}

				// bannedlist
				if ( empty( $input['bannedlist'] ) ) {
					update_site_option( 'banhammer_keys', '' );
				} elseif ( $input['bannedlist'] !== $this->bannedlist ) {
					$new_bannedlist = explode( "\n", $input['bannedlist'] );
					$new_bannedlist = array_filter( array_map( 'trim', $new_bannedlist ) );
					$new_bannedlist = array_unique( $new_bannedlist );
					foreach ( $new_bannedlist as &$keyname ) {
						$keyname = sanitize_text_field( $keyname );
					}
					$new_bannedlist = implode( "\n", $new_bannedlist );
					update_site_option( 'banhammer_keys', $new_bannedlist );
				}
				unset( $input['bannedlist'] );

				// Redirect
				if ( ! isset( $input['redirect'] ) || is_null( $input['redirect'] ) || '0' === $input['redirect'] ) {
					$output['redirect'] = false;
				} else {
					$output['redirect'] = true;
				}

				// Redirect URL
				if ( isset( $input['redirect_url'] ) && $input['redirect_url'] !== $options['redirect_url'] ) {
					$output['redirect_url'] = esc_url( $input['redirect_url'] );
				} else {
					$output['redirect_url'] = $options['redirect_url'];
				}

				// User name / login
				if ( ! isset( $input['check_user_login'] ) || is_null( $input['check_user_login'] ) || '0' === $input['check_user_login'] ) {
					$output['check_user_login'] = false;
				} else {
					$output['check_user_login'] = true;
				}

				$this->options = $output;
				update_site_option( $this->option_name, $output );

				?>
				<div class="notice notice-success is-dismissible"><p><strong><?php esc_html_e( 'Options Updated!', 'ban-hammer' ); ?></strong></p></div>
				<?php
			}
		}
	}

	/**
	 * Register Admin Settings
	 *
	 * @since 2.6
	 */
	public function register_settings() {
		register_setting( 'ban-hammer', 'banhammer_options', array( &$this, 'banhammer_sanitize' ) );

		// The main section
		add_settings_section( 'banhammer-settings', '', array( &$this, 'banhammer_settings_callback' ), 'ban-hammer-settings' );

		// The Fields
		add_settings_field( 'message', __( 'Blocked Message', 'ban-hammer' ), array( &$this, 'message_callback' ), 'ban-hammer-settings', 'banhammer-settings' );
		add_settings_field( 'redirect', __( 'Redirect Blocked Users?', 'ban-hammer' ), array( &$this, 'redirect_callback' ), 'ban-hammer-settings', 'banhammer-settings' );
		add_settings_field( 'bannedlist', __( 'The Blocked List', 'ban-hammer' ), array( &$this, 'bannedlist_callback' ), 'ban-hammer-settings', 'banhammer-settings' );
		add_settings_field( 'check_user_login', __( 'Check User Login?', 'ban-hammer' ), array( &$this, 'check_user_login_callback' ), 'ban-hammer-settings', 'banhammer-settings' );
	}

	/**
	 * Ban Hammer Settings Callback
	 *
	 * @since 2.6
	 */
	public function banhammer_settings_callback() {
		?>
		<p><?php esc_html_e( 'Customize your Ban Hammer experience via the settings below.', 'ban-hammer' ); ?></p>
		<?php
	}

	/**
	 * The Banned List Callback
	 *
	 * @since 2.6
	 */
	public function bannedlist_callback() {
		?>
		<p><?php esc_html_e( 'The terms below will not be allowed to be used during registration. You can add in full emails (i.e. foo@example.com) or domains (i.e. @domain.com), and partials (i.e. viagra). Wildcards (i.e. *) will not work.', 'ban-hammer' ); ?></p>
		<p><textarea name="banhammer_options[bannedlist]" id="banhammer_options[bannedlist]" cols="40" rows="15"><?php echo esc_textarea( $this->bannedlist ); ?></textarea></p>
		<?php
	}

	/**
	 * Message Callback
	 *
	 * @since 2.6
	 */
	public function message_callback() {
		?>
		<p><?php esc_html_e( 'The message below is displayed to users who are not allowed to register on your blog. Edit is as you see fit, but remember you don\'t get a lot of space so keep it simple.', 'ban-hammer' ); ?></p>
		<p><textarea name="banhammer_options[message]" id="banhammer_options[message]" cols="80" rows="2"><?php echo esc_html( $this->options['message'] ); ?></textarea></p>
		<?php
	}

	/**
	 * Redirect Callback
	 *
	 * @since 2.6
	 */
	public function redirect_callback() {
		?>
		<p><?php esc_html_e( 'If you\'d rather redirect users to a custom URL, please check the box below. If you do, the message above will not show.', 'ban-hammer' ); ?></p>
		<p><input type="checkbox" id="banhammer_options[redirect]" name="banhammer_options[redirect]" value="yes" <?php checked( $this->options['redirect'], 'yes', true ); ?> <?php checked( $this->options['redirect'], '1', true ); ?> >
		<label for="banhammer_options[redirect]"><?php esc_html_e( 'Redirect failed logins to a custom URL.', 'ban-hammer' ); ?></label></p>

		<?php
		if ( isset( $this->options['redirect'] ) && 'no' !== $this->options['redirect'] && ! empty( $this->options['redirect'] ) ) {
			?>
			<p><textarea name="banhammer_options[redirect_url]" id="banhammer_options[redirect_url]" cols="60" rows="1"><?php echo esc_url( $this->options['redirect_url'] ); ?></textarea>
			<br /><span class="description"><?php esc_html_e( 'Set redirect URL (example: http://example.com).', 'ban-hammer' ); ?></span></p>
			<?php
		}
	}
	
	/**
	 * Check User Login Callback
	 *
	 * @since 2.9
	 */
	public function check_user_login_callback() {
		?>
		<p><?php esc_html_e( 'In addition to checking email addresses against the blocked list, you can also check the user name / login ID.', 'ban-hammer' ); ?></p>
		<p><input type="checkbox" id="banhammer_options[check_user_login]" name="banhammer_options[check_user_login]" value="no" <?php checked( $this->options['check_user_login'], 'yes', true ); ?> <?php checked( $this->options['check_user_login'], '1', true ); ?> >
		<label for="banhammer_options[check_user_login]"><?php esc_html_e( 'Test user name in registration.', 'ban-hammer' ); ?></label></p>

		<?php
	}

	/**
	 * Options
	 *
	 * The options page. Since this has to run on Multisite, it can't use the settings API.
	 *
	 * @since 1.0
	 * @access public
	 */
	public function options() {
		?>
		<div class="wrap">

		<h1><?php esc_html_e( 'Ban Hammer', 'ban-hammer' ); ?></h1>

		<?php
		settings_errors();

		if ( is_network_admin() ) {
			?>
			<form method="post" width='1'>
			<?php
			wp_nonce_field( 'banhammer_networksave' );
		} else {
			?>
			<form action="options.php" method="POST" >
			<?php
			settings_fields( 'ban-hammer' );
		}
				do_settings_sections( 'ban-hammer-settings' );
				submit_button( '', 'primary', 'update' );
		?>
			</form>
		</div>
		<?php
	}

	/**
	 * Options sanitization and validation
	 *
	 * @param $input the input to be sanitized
	 * @since 2.6
	 */
	public function banhammer_sanitize( $input ) {
		// Get current options
		$options = $this->options;

		// Hardcoding becuase it's always this.
		$input['db_version'] = $this->db_version;

		// Message
		if ( $input['message'] !== $options['message'] && wp_kses_post( $input['message'] === $input['message'] ) ) {
			$input['message'] = wp_kses_post( $input['message'] );
		} else {
			$input['message'] = $options['message'];
		}

		// bannedlist
		if ( empty( $input['bannedlist'] ) ) {
			update_option( 'disallowed_keys', '' );
		} elseif ( $input['bannedlist'] !== $this->bannedlist ) {
			$new_bannedlist = explode( "\n", $input['bannedlist'] );
			$new_bannedlist = array_filter( array_map( 'trim', $new_bannedlist ) );
			$new_bannedlist = array_unique( $new_bannedlist );
			foreach ( $new_bannedlist as &$keyname ) {
				$keyname = sanitize_text_field( $keyname );
			}
			$new_bannedlist = implode( "\n", $new_bannedlist );
			update_option( 'disallowed_keys', $new_bannedlist );
		}
		unset( $input['bannedlist'] );

		// Redirect
		if ( ! isset( $input['redirect'] ) || is_null( $input['redirect'] ) || '0' === $input['redirect'] ) {
			$input['redirect'] = 'no';
		} else {
			$input['redirect'] = 'yes';
		}

		// Redirect URL
		if ( isset( $input['redirect_url'] ) && $input['redirect_url'] !== $options['redirect_url'] ) {
			$input['redirect_url'] = esc_url( $input['redirect_url'] );
		} else {
			$input['redirect_url'] = $options['redirect_url'];
		}
		
		// User login / name
		if ( ! isset( $input['check_user_login'] ) || is_null( $input['check_user_login'] ) || '0' === $input['check_user_login'] ) {
			$input['check_user_login'] = 'no';
		} else {
			$input['check_user_login'] = 'yes';
		}

		return $input;
	}

	/**
	 * Banhammer
	 *
	 * Here's the basic plugin for WordPress SANS BuddyPress
	 *
	 * @since 1.0
	 * @access public
	 */
	public function banhammer_drop( $user_login, $user_email, $errors ) {
		$bannedlist_string = $this->bannedlist;
		$bannedlist_array  = explode( "\n", $bannedlist_string );

		$drop = $this->test_drop($bannedlist_array, $user_email, $errors);
		if($drop)
			return $drop;

		if($this->options('check_user_login') ?? false) {
			$drop = $this->test_drop($bannedlist_array, $user_login, $errors);
		}
		if($drop)
			return $drop; // yes, redundant at this moment, soon not...

		return $drop;
	}

	/**
	 * Actual testing done here
	 *
	 * @since 2.9
	 * @access private
	 */
	private function test_drop($bannedlist_array, $field, $errors) {
		$bannedlist_size   = count( $bannedlist_array );
		// Go through bannedlist
		for ( $i = 0; $i < $bannedlist_size; $i++ ) {
			$bannedlist_current = trim( $bannedlist_array[ $i ] );
			if ( stripos( $field, $bannedlist_current ) !== false ) {

				$errors->add( 'invalid_email', $this->options['message'] );
				if ( 'yes' === $this->options['redirect'] ) {
					wp_safe_redirect( $this->options['redirect_url'] );
				} else {
					return true;
				}
			}
		}
	}


	/**
	 * Validation
	 *
	 * Validate the keys for Multisite
	 *
	 * @since 1.0
	 * @access public
	 */
	public function wpmu_validation( $result ) {
		if ( $this->banhammer_drop( sanitize_user( $_POST['user_name'] ), sanitize_email( $_POST['user_email'] ), $result['errors'] ) ) {
			$result['errors']->add( 'user_email', $this->options['message'] );
		}
		return $result;
	}

	/**
	 * Check if registration is active
	 *
	 * @since 2.8
	 * @access public
	 */
	public function is_registration_active() {

		// List of functions for plugins that don't really use the registration
		// form, default to WP:
		$has_special_plugin = false;
		$special_plugins    = array(
			'is_woocommerce_activated',      // WooCommerce
			'wpforo_load_plugin_textdomain', // WPForo
			'gf_user_registration',          // Gravity Forms: User Registration

		);

		// Loop through the functions, if they exist, we skip.
		foreach ( $special_plugins as $this_plugin ) {
			if ( ! $has_special_plugin && function_exists( $this_plugin ) ) {
				$has_special_plugin = true;
			}
		}

		// Filter so other people can get around this.
		// If you want to skip it, it's:
		// add_filter( 'ban_hammer_alt_registration', true );
		$alt_registration = apply_filters( 'ban_hammer_alt_registration', $has_special_plugin );

		// Warn if Registration isn't active or certain plugins are active
		if ( ! get_option( 'users_can_register' ) && user_can( get_current_user_id(), 'moderate_comments' ) && ! $alt_registration ) {
			$this->notice_message = __( 'Ban Hammer requires standard WordPress registration to be enabled.', 'ban-hammer' );
			$this->notice_class   = 'warning';

			if ( is_multisite() ) {
				add_action( 'network_admin_notices', array( &$this, 'admin_notices' ), 10, 2 );
			} else {
				add_action( 'admin_notices', array( &$this, 'admin_notices' ), 10, 2 );
			}
		}
	}

	/**
	 * Donate link
	 *
	 * Adds link to donate on the plugins page
	 *
	 * @access public
	 */
	public function donate_link( $links, $file ) {
		if ( plugin_basename( __FILE__ ) === $file ) {
				$donate_link = '<a href="https://ko-fi.com/A236CEN/">' . __( 'Donate', 'ban-hammer' ) . '</a>';
				$links[]     = $donate_link;
		}
		return $links;
	}

	/**
	 * Settings link
	 *
	 * Adds link to settings page on the plugins page
	 *
	 * @access public
	 */
	public function settings_link( $links ) {
		if ( is_multisite() ) {
			$settings_link = network_admin_url( 'settings.php?page=ban-hammer' );
		} else {
			$settings_link = admin_url( 'tools.php?page=ban-hammer' );
		}

		$settings_link = '<a href="' . $settings_link . '">' . __( 'Settings', 'ban-hammer' ) . '</a>';
		if ( user_can( get_current_user_id(), 'moderate_comments' ) ) {
			array_unshift( $links, $settings_link );
		}

		return $links;
	}

}

new BanHammer();
